grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +175,122 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
         return plan;
     }
 
+    private static void inferRelationshipTypes(EmbeddedGraknTx<?> tx, Set<Fragment> allFragments) {
+        // find all vars representing types
+        Map<Var, Type> labelVarTypeMap = new HashMap<>();
+        allFragments.stream()
+                .filter(LabelFragment.class::isInstance)
+                .forEach(fragment -> {
+                    // TODO: labels() should return ONE label instead of a set
+                    Type type = tx.getType(Iterators.getOnlyElement(((LabelFragment) fragment).labels().iterator()));
```

Comment:
is it possible to get rules here?

Reply:
I think so. You can query for rules like `match $x label grandparent-rule; get` or whatever.

But I think I was wrong about it crashing! If you call `tx.getType("grandparent-rule");`, it will return `null` and you handle that.